### DISCUSSION TITLE
Context Menu items are disabled when binding to a collection

### DISCUSSION THREAD

**Discussion by ClassyCircuit** [AvaloniaUI/Avalonia#14591] [Q&A] [answered]
Context Menu items are disabled when binding to a model and initializing collection during runtime:

This doesn't work, item is disabled:
```axaml
<TextBlock VerticalAlignment="Center" Text="{Binding Name}">
	<TextBlock.ContextMenu>
		<ContextMenu ItemsSource="{Binding ContextMenuItems}">
			<ContextMenu.ItemTemplate>
				<DataTemplate DataType="vm:ContextMenuItemModel">
					<MenuItem Header="{Binding Header}" Command="{Binding Cmd}"/>
				</DataTemplate>
			</ContextMenu.ItemTemplate>
		</ContextMenu>
	</TextBlock.ContextMenu>
</TextBlock>
```
![image](https://github.com/AvaloniaUI/Avalonia/assets/6424248/6a82f82a-a3e7-41c0-8af3-d4661a3a9785)

```C#
    /// Class for context menu items
    public class ContextMenuItemModel
    {
        public string Header { get; set; }
        public ICommand Cmd { get; set; }

        public ContextMenuItemModel(string header, ICommand cmd)
        {
            Header = header;
            Cmd = cmd;
        }
    }

   /// Class that initializes the menu
   public class Block
   {
        public List<ContextMenuItemModel> ContextMenuItems { get; } = [];
        public Block()
        {
                ContextMenuItems.Add(new ContextMenuItemModel("New Request", new RelayCommand(CreateNewRequestAsync)));
        }

        public void CreateNewRequestAsync()
        {
            System.Diagnostics.Trace.WriteLine($"Test");
        }
   }
```

This works (item is not disabled):
```axaml
<TextBlock VerticalAlignment="Center" Text="{Binding Name}">
	<TextBlock.ContextMenu>
		<ContextMenu>
			<MenuItem Header="Open" Command="{Binding CreateNewRequestAsyncCommand}" />
		</ContextMenu
	</TextBlock.ContextMenu>
</TextBlock>
```

![image](https://github.com/AvaloniaUI/Avalonia/assets/6424248/4353b38f-5f25-4554-a993-f225b14662b4)

What is happening here, why can't I initialize the context menu during runtime by binding to a model?

## Environment 
- OS: Windows 11
- Avalonia-Version: 11.0.9

**Comment by StefanKoell** [comment]
If I'm not mistaken you need to implement CanExecute and return true in order to enable the command:
https://learn.microsoft.com/en-gb/dotnet/api/system.windows.input.icommand.canexecute?view=netstandard-2.0#system-windows-input-icommand-canexecute(system-object)

Did you try that?

**Reply by ClassyCircuit** [reply]
Yes, I tried that, but it didn't change anything.

**Comment by timunie** [comment]
You shouldn't use a MenuItem inside DataTemplate. Use a Style bound to the command. 

Note: Please use Q&A for questions, not bug reports.

**Reply by ClassyCircuit** [reply]
Thanks, I will try that. Is there some documentation on how to implement that?
Sorry for creating a bug report.

**Reply by timunie** [reply]
File a minimal sample pls

**Comment by gehongyan** [comment, marked as answer]
> Thanks, I will try that. Is there some documentation on how to implement that? Sorry for creating a bug report.

See the sample:

https://github.com/AvaloniaUI/Avalonia/blob/a33b39e8e021e01290d8d714f51ae17752fb03fd/samples/ControlCatalog/Pages/ContextMenuPage.xaml#L50-L63

![image](https://github.com/AvaloniaUI/Avalonia/assets/21241496/2de1722c-973a-49ca-a8bf-ebbd3d93fdd4)

**Reply by ClassyCircuit** [reply]
Thank you for the sample, it works perfectly.
